### ISSUE TITLE
[Feature] `azd env delete` command to delete an existing environment

### ISSUE THREAD

**Issue by philliphoff** [Azure/azure-dev#937]
While a `azd env new` command exists to create environments, there's no symmetric command for deleting them.  While users/tooling can remove the environment folder manually, there's related AZD internal state (such as the current default environment setting) that may be impacted and would be best for the `azd` CLI to manage.

**Comment by savannahostrowski** [comment]
This seems like something that we should implement for consistency @rajeshkamal5050

**Comment by bwateratmsft** [comment]
This is required for #1554.

**Comment by weikanglim** [comment]
Naming: I quite like @ellismg suggestion of `azd env rm`. Keeping consistent with our command structure, I'm guessing we might have `azd env unset` instead to mean "remove values".

Feature possibilities of `env rm/delete`:
- If environment is provisioned (has `AZURE_SUBSCRIPTION_ID`), provide message + command for cleanup (would need `azd down` to accept something like a ARM deployment ID)? Alternatively, is to block/prompt cleanup.
     - The cost of any checks of "live resources remaining" is currently at least 10s (bicep compile + ARM deployment fetch), excluding additional resource fetches.
     - Note the interesting design here: Only with environment name/sub/location + bicep file (scope) can we recover the resources. If user changed the bicep from subscription -> RG, it destroys the link from environment to resources.
- If environment has remote store reference, prompt for remote deletion? @wbreza 
- If environment is set as default environment, unset the default environment. My preference is to delay the asking of "new default" until the next command (see #3357)
- We may need some sort of `-f` or `--force` to bypass some confirmations (once we solidify the above)

I think @vhvb1989 also raised the possibility of "soft-deletion" where environments could be recoverable. Technically with the deployment state out there, it is possible to recover by recreating an env with the same IDs (name/sub/location). We'd need to think also about env config. The main thing I'd think users would want is that sensitive data needs to be purged.

I suspect if we're doing anything novel here, we'd want to make the link between environment to Azure resources more concrete.

**Comment by rajeshkamal5050** [comment]
Getting implemented as part of #3298. Marking this as a duplicate.

**Comment by weikanglim** [comment]
Going to keep this open just to have the CLI feature discussion separate from VS integration.

**Comment by weikanglim** [comment]
Noting some UX discussion items:

1. Naming: `env rm` vs. `env delete`? My vote is `env rm` currently. `rm` is shorter; "remove" as a more positive tonal feeling than "delete".
2. What happens when user runs `azd env rm` before `azd down`: Currently, there isn't a way to trace back to the resources/deployment when the `.env` file is removed. This means that there isn't an azd gesture to run `azd down` after `azd env rm`. Expectation would be that user needs to manually find the RGs and delete.
    - Long-term I believe we should support this, but as a follow-up feature when `azd down` is more intelligent (Azure Stacks).
4. If we do require manual clean-up on 2, I think we want to be extremely clear in the wording. I've chosen a "softer" title-notes approach to start with:
<img width="1512" alt="image" src="https://github.com/Azure/azure-dev/assets/2322434/9a5e7aa9-cc4f-4d04-858b-195c21c6c806">

**Comment by rajeshkamal5050** [comment]
1. `rm` feels short form how about `remove` aligned with other commands we have in full form?
2. If resources aren't removed via `azd down`. Can't `azd env remove` validate and fail `Azure resources are still associated with 'acme' environment. Delete Azure resources using `azd down` before running `azd env remove`. This could help with avoiding manual clean-ups.

**Comment by philliphoff** [comment]
My $0.02: `rm` makes assumptions about user's knowledge of a particular OS, without which they may not understand its purpose.  I'd vote for `remove` or `delete` so as to be entirely clear.

**Comment by weikanglim** [comment]
> rm feels short form. How about remove sounds aligning with other commands we have in full form?

We have `init`, `config`, `synth`, `env` currently. 

> If resources aren't removed via azd down. Can't azd env remove validate and fail Azure resources are still associated with 'acme' environment.

We can. Mentioned this above, but the cost of any checks of "live resources remaining" is currently at least 10s (bicep compile + ARM deployment fetch), excluding additional resource fetches.

There's also an orthogonality factor here between "env resources" and "Azure resources" -- there could be reasons to delete the environment (remove values storage), but keep the Azure resources live.

**Comment by wbreza** [comment]
If we are only removing the environment folder `azd env remove <name>` (Can include an `rm` alias) sounds like a good fit.  

If we plan to additionally performing an `azd down` or similar and deleting resources then I feel like `azd env delete` may be better since it completely removing the deployed resources plus environment vars/config.

**Feedback:**
- If providing a prompt on whether or not to perform the delete on the remote environment store I would additionally add an additional flag like `--force`, `--remote`, `--all` or similar to avoid these prompts.  That also raises the question on what is the default behavior if someone forced this?
- Additional context when platform type is `devcenter`.  Deleting a remote devcenter environment would perform a resource deletion.

**Comment by weikanglim** [comment]
> If providing a prompt on whether or not to perform the delete on the remote environment store I would additionally add an additional flag like --force, --remote, --all or similar to avoid these prompts. That also raises the question on what is the default behavior if someone forced this?

My plan is to have `--force` mean "delete all", we can slowly expand for `--remote` or `--all` in the future.

> Additional context when platform type is devcenter. Deleting a remote devcenter environment would perform a resource deletion.

Does that mean that in `devcenter` mode, `azd env remove` is an alias for `azd down`? Because I believe `azd down` does this currently.

**Comment by wbreza** [comment]
@weikanglim 

> Does that mean that in devcenter mode, azd env remove is an alias for azd down? Because I believe azd down does this currently.

No, nothing like that.  Just that calling the ADE delete environment API internally deletes the resource group similar to our `azd down` command.

**Comment by ellismg** [comment]
> No, nothing like that. Just that calling the ADE delete environment API internally deletes the resource group similar to our azd down command.

That's sort of unfortunate - but I guess that is the reality of the ADE API we have to interact with.  Taking ADE out of the question - it does feel like the default of `azd env rm` should be to remove the local and remote copy of the environment (perhaps with bool flags like `--local` or `--remote` to control this behavior).  I feel like this is what you would want as the default for cases where you are say using blob storage for you environment, otherwise calls like this:

```
$ azd env new testing
$ azd env set foo bar
$ azd env rm
$ azd env new testing # <--- this should work without any name in use issues.
$ azd env get-values # <--- this should not have foo=bar as set previously, since we remove the environment 
```

May not behave as people expect?  I would expect that I don't see values from the old environment.

For the ADE side we might need to add a prompt that says "removing this environment will also remove any resources associated with it (and maybe a count of the number of resources, like we do in down, and only show this when the count is non zero)" and `--force` lets you skip past that prompt?

**Comment by weikanglim** [comment]
> For the ADE side we might need to add a prompt that says "removing this environment will also remove any resources associated with it (and maybe a count of the number of resources, like we do in down, and only show this when the count is non zero)" and --force lets you skip past that prompt?

Just for my own understanding, if we're going down this route, when would we advise a  devcenter user to run `azd down` vs. `azd env rm`? My understanding is that `azd down`  in devcenter mode does precisely this, see [Destroy](https://github.com/Azure/azure-dev/blob/ab9976a35dc8462299334dd86f129318378a450a/cli/azd/pkg/devcenter/provision_provider.go#L250).

**Comment by mindlessroman** [comment]
Would love this ! (Pinging to revive attention)

**Comment by l3ender** [comment]
Until a command exists to delete an environment, is there a file or some other configuration that can be deleted to remove one?

**Comment by vhvb1989** [comment]
> Until a command exists to delete an environment, is there a file or some other configuration that can be deleted to remove one?

@l3ender , you can manually delete the environment folder from `.azure` folder.  There is one folder with the name of each environment in there. Just make sure you have called `azd down` or manually delete your resources.

**Comment by l3ender** [comment]
> you can manually delete the environment folder from `.azure` folder. There is one folder with the name of each environment in there. Just make sure you have called `azd down` or manually delete your resources.

Thank you! I was looking in `~/.azure` (home directory) instead of the `.azure` folder in the project directory. I see it now--thank you.

**Comment by vhvb1989** [comment]
`~/.azure` is owned by `az cli`
`~/.azd` is where you would find azd's user config (for all projects)
`.azure` within each project is where each env-folder name is kept.

**Comment by spboyer** [comment]
when calling `azd env delete <env name>` prompt the user through the `azd down` option > if yes then `down` if no then just delete the ./azure/<env name>

@weikanglim - can we get this completed ?

**Comment by vhvb1989** [comment]
> when calling `azd env delete <env name>` prompt the user through the `azd down` option > if yes then `down` if no then just delete the ./azure/
> 
> [@weikanglim](https://github.com/weikanglim) - can we get this completed ?

It could be a great candidate for an azd-extension (for environments) which could handle
 -  soft-delete 
 -  copy/clone
 -  export/import
 -  versioning/snapshoots
 -  etc.
